### ISSUE TITLE
Query engine library for current platform "windows" could not be found

### ISSUE THREAD

**Issue by leegee** [prisma/prisma#10714]
Suddenly, for no apparent reason:

```
    Query engine library for current platform "windows" could not be found.
    You incorrectly pinned it to windows

    This probably happens, because you built Prisma Client on a different platform.
    (Prisma Client looked in "L:\src\foaf\node_modules\@prisma\client\runtime\query_engine-windows.dll.node")

    Searched Locations:

      L:\src\foaf\node_modules\.prisma\client
      L:\src\foaf\node_modules\@prisma\client
      L:\src\foaf\node_modules\@prisma\client
      L:\src\foaf\node_modules\.prisma\client
      L:\src\foaf\prisma
      /tmp/prisma-engines
      L:\src\foaf\node_modules\.prisma\client

    You already added the platform "windows" to the "generator" block
    in the "schema.prisma" file as described in https://pris.ly/d/client-generator,
    but something went wrong. That's suboptimal.

    Please create an issue at https://github.com/prisma/prisma/issues/new

      at Object.getLibQueryEnginePath (node_modules/@prisma/client/runtime/index.js:36461:13)
      at Object.loadEngine (node_modules/@prisma/client/runtime/index.js:36129:33)
      at Object.instantiateLibrary (node_modules/@prisma/client/runtime/index.js:36095:5)

$ grep 'prisma'  package.json
    "prisma": "^3.6.0",

$ node -v
v16.10.0

$ npm -v
7.24.1

```
Shell is git-bash.

What can I tell you?

**Comment by janpio** [comment]
> What can I tell you?

Define "suddenly" (no change at all? update? next day? new installation of the project?).
Fill the bug report issue template with the information it asks for.

**Comment by leegee** [comment]
I followed the instruction the library gave me, as quoted above. There was no 'issue template'.

"Define "suddenly""?  As in, it happened quickly and without warning. No, I didn't notice doing anything.

**Comment by bhuddleston-biocorellc** [comment]
@leegee were you able to resolve this? If so, would you mind detailing how you did? I'm running into the same issue. Thank you.

**Comment by appurist** [comment]
Jumping in here because I think I just resolved my case of this.

Are you using `esbuild`, or any build process that copies files to a target build folder?  I was using esbuild and the output folder was .esbuild and my particular case shows a path relative to that, after the `Prisma Client looked in ...` error. So it was looking for the dll in a specific place it seems and not finding it (copied) there. I manually copied it to `.esbuild\.build\node_modules\.prisma\client\` where the error indicated and now it is able to make queries.

If that "Prisma Client looked in..." path looks correct, and the specified windows DLL file is there, then it may be that it was built on a different machine and not the correct one (or 32-bit vs 64-bit node, or some other difference from the current setup. 

In that case, try verifying the path in the "looked in" message exists, and if it does, nuke your `node_modules` folder and make sure `@prisma/client` and `prisma` are the same version in the `package.json` file.  Rerun `npm ci` for a clean install, run `npx prisma generate` and try starting your node app again. That worked for me, after I copied the windows DLL file from `node_modules\.prisma\client\` to the place mentioned in the "looked in" message.

**Comment by janpio** [comment]
@appurist That is some nice investigation and understanding. Can you maybe open a new issue (bug report) about this default behavior? Optimally none of this would be necessary of course!

**Comment by appurist** [comment]
@janpio I'm not sure yet if there's actually a bug, unless `prisma generate` is missing a copy or hard-coding a path. I haven't had time to follow-up yet on why our automation didn't copy the engine .node file over, because I found only the linux one on our Windows build, and I don't yet see where we have that assumption. So maybe it is in prisma, but I won't assume it's a bug there yet. I only posted this here because I think this is just the general pattern of the "looked in" message giving us a solution if the file isn't actually there and one was built/available. If I find it's not our CI/CD process, I'll file the bug report as suggested. It may be a few days though, other urgent issues to resolve in my project.

**Comment by janpio** [comment]
That would be nice. I can totally believe this is a bug, because we know that a) the message logic is super wonky and b) using bundlers with our current Prisma Client code has many ways to mess things like this up. Especially when `binaryTargets` are used to get multiple engines for different platforms, or using a custom `output` folder for the `generator client` in the schema. (One of those apply to you by chance?)

**Comment by appurist** [comment]
@janpio Okay, the problem seemed to just be that I was using esbuild to produce the output folder and the includes/excludes were copying the linux version only. Our build is eventually for AWS Lambda, built from a Docker image has a different distro, so we specifically build the "rhel-openssl-1.0.x" version for AWS Lambda. But none of that is too relevant here, I just mention it because you asked about multiple targets:
```
generator client {
  provider      = "prisma-client-js"
  binaryTargets = ["native", "rhel-openssl-1.0.x"]
}
```
This problem was on Windows and our `.esbuild` didn't include the `.node` engine file for Windows, because it was excluded in the serverless.yml file. I added a `*.node` wildcard instead of the rhel-specific path that was there and it's fine.
```
package:
  excludeDevDependencies: false
  patterns:
    - '!node_modules/.prisma/client/libquery_engine-*'
    - '!node_modules/prisma/libquery_engine-*'
    - '!node_modules/@prisma/engines/**'
    - 'node_modules/.prisma/client/schema.prisma'
    - 'node_modules/.prisma/client/*.node'
```
That last line ensures these are included. We originally had a problem with the size of prisma in the Lambda image so we started filtering down files that weren't needed at runtime. I don't think the others are still needed but the last line ensures we have the native engine, whatever that is.
